### ISSUE TITLE
Invalid JSON reported in Routing-Rule editor for property replacement variables

### ISSUE THREAD

**Issue by Kusig** [swisspush/gateleen#382]
If there is used  property replacement variable for a numeric JSON attribute, the REST Editor does mark this as invalid JSON. 

Example:
```
(/myserver/services)?/mylink(/$)?(.*)": {
    "description": "blabla",
    "url": "${my.link}$3",
    "proxyOptions": {
      "type": "HTTP",
      "host": "${outproxy.link}",
      "port": ${outproxy.port}
    }
  }
```

The error is basically right because this is indeed invalid JSON at this time because the property-variable will be replaced only at runtime.

Reference issue: https://jira.post.ch/browse/SDCISA-1202

**Comment by Kusig** [comment]
In oder do define such objects always as valid JSON, I propose to define the numeric attribute with a slightly different syntax:

"port": "$${outproxy.link}"

This way we could easily change Gateleen StringUtils  to replace this special tag without double quotes resulting in:

"port": 1234

This would work the same way just out of the box for boolean attributes as well .  It is 100% backwards compatible without any major risks.

@lbovet  WDYT?

**Comment by lbovet** [comment]
An alternative would be to accept also strings for the port. It is cheaper, we don't add a feature to implement.

**Comment by Kusig** [comment]
In this case we would have to create our own Proxy Class in between because the current implementation directly uses the vert.x class which we can not change.  As well, it would only work for this dedicated case and not in general.  As the feature looks to be very simple to be implemented, I'll go on with that one and will create a PR about.

**Comment by lbovet** [comment]
Or simply transform the value here: https://github.com/swisspush/gateleen/blob/develop/gateleen-routing/src/main/java/org/swisspush/gateleen/routing/RuleFactory.java#L142
We should be able to keep a decoupling between our configuration API and the internal classes we use from vert.x

**Comment by Kusig** [comment]
That's where I was already but I am not happy with it and discardet it.

This would only resolve exactly and only this case and only for this object and ruleset. It would have to be implemented for any other case with the same problem (eg. already just a boolean would cause the same problem).

From my point of view we need a generic approach where we could signal the system to transform the attribute to the right JSON type before the JSON transformation to the corresponding object is executed.

**Comment by lbovet** [comment]
If we want the cleanest solution, we should address the problem where it is, i.e. in the REST editor to make it aware of variables, (and also able to check the that the variable language is correct). I think there are possibility to enable multiple syntaxes with the ace editor.

**Comment by mcweba** [comment]
I agree with the comment from @lbovet. The problem is only in the REST editor, since we have json schema validation (i.e. for rules) which is validated against **after** the wildcards have been replaced.

**Comment by Kusig** [comment]
Yes and No I would say. The JSON is not valid at this point. Therefore any extension there must make it aware of the wildcard syntax and break the JSON schema therefore....  Is this worth the effort and not much more risky than simply give us the power of configuring and controlling what must be replaced. This makes it much more visible then to any reviewers of the rules and it doesn't hide all the magic somewhere.  
I personally prefere to keep the strict JSON format and to enforce this one and to explicitely configure the type of the wildcard. Allowing non valid JSON for Wildcards is more risky than defining explicitely the type conversion.

"port" : @{myport}    -> invalid JSON to be tweaked by schema and editor somehow under the hoods. 
    vs.
"port": "@@{myport}"  -> valid JSON and it is clear what will happen here by definition


See my PR for this.

**Comment by lbovet** [comment]
The proposed fix addresses only one case. The templating engine (jmte) has other features like conditionals and loops. They would not be covered by this fix and not be displayed correctly in the rest editor anyway.
That said, this is a quick win for the most frequent cases (non-string values) and it would also avoid other editors (vscode) to complain.
This problem is ubiquitous with non-structural templating languages.
Maybe we should switch to YAML and [Yglu](https://yglu.io/) :)

**Comment by Kusig** [comment]
I do not agree on that. The fix allows us to use proper JSON in any case. That's exactly the reason why I proposed this one, it makes as well any other editor happy at the same time and this is important. The fix shows therefore properly in the editor and works in any case where the existing replacement already does, it's just a second one with clear deterministic behaviour. 

Fixing it in the Gateleens editor itself would lead to ignored lines with parameters and therefore hide any other problem on this line. And as mentioned before, it would anyhow not work with our real used editors outside. 

Therefore I stick with my proposaed PR.

**Comment by lbovet** [comment]
I meant, the proposed fix is a quick win that covers only the most frequent case where we have non-string values, but it will not help if we use other templating features (if and conditionals). It is not the cleanest solution but there is probably no better alternative since templating is not generally supported in editors. So, for me it is okay-ish and acceptable. It addresses a particular need as a kind of workaround for a problem that has no general solution. The argument about support in other editors is determinant for me, in this case. I let the others come with a better alternative if they have one...

**Comment by lbovet** [comment]
Still, my preferred approach would be to accept strings everywhere, since the source are strings. It is simpler and does not introduce an additional syntax.

**Comment by mcweba** [comment]
Released in [v1.1.73](https://github.com/swisspush/gateleen/releases/tag/v1.1.73)
